### ISSUE TITLE
About PowerPC Support

### ISSUE THREAD

**Issue by jpc0016** [JonathanSalwan/Triton#832]
Hi Jonathan,

Based on Triton's dependency on Pintool, I'm wondering if PowerPC support can be added to the architectures list.  If possible, which files need to be altered/added to ensure support?

**Comment by JonathanSalwan** [comment]
> Based on Triton's dependency on Pintool,

What do you mean? Pintool is not a dependency of Triton but if you want to use our pintool tracer, you can not have the PowerPC support. Pintool only support x86.

**Comment by jpc0016** [comment]
Sorry for the confusion.  ARM support was added to Triton following discussion in #420 .  I would like to do the same for PowerPC.  So far I need to add one ppc folder in /arch/, and add four files: ppc64cpu.cpp, ppccpu.cpp, ppcsemantics.cpp, ppcspecifications.cpp.  From your response, changes are also needed in `SymbolicEngine.cpp`.  What other changes are required?

**Comment by JonathanSalwan** [comment]
We did not add the ARM support during #420, but we do now (#831). Basically, in order to add a new architecture you have to:

1) Create a new directory in `src/libtriton/arch/`
2) Add files : `archCpucpp`, `aarchSemantics.cpp` and `archSpecifications.cpp`  ([skeleton](https://github.com/JonathanSalwan/Triton/tree/master/src/libtriton/arch/aarch64))
3) Add the `arch.spec` file in `src/includes` ([skeleton](https://github.com/JonathanSalwan/Triton/blob/master/src/libtriton/includes/triton/aarch64.spec))
4) Update core base files where `ARCH_*` are used in order to add your new enum.
5) Make unittests to test your arch semantics ([skelaton](https://github.com/JonathanSalwan/Triton/blob/master/src/testers/unicorn_test_aarch64.py))

Be careful, don't touch `SymbolicEngine.cpp` and other core files because we are currently adding the ARM support and if you do the same for PPC we will probably have conflicts...

For example, https://github.com/JonathanSalwan/Triton/commit/349da6be947169ad88d3d14e11e4477416e0e706 is the commit  (which may help you) where the AArch64 has been added.

Question : What is the endianness of your PPC support ? If it is little endian, it will be easy to implement the support but if it is in big endian, we have to modify the `SymbolicEngine.cpp` and probably others files.

**Comment by jpc0016** [comment]
Thanks Jonathan!  It is 32-bit PPC big endian so yes, additional changes are necessary.

**Comment by jpc0016** [comment]
I forked Triton's master repository and added PPC32 skeleton files based on commit [d1b9817](https://github.com/JonathanSalwan/Triton/commit/ab37265adf9e5b9ede67a0c01d8bf7366d1b8917).  I'm sort of following along each commit in thread [#831](https://github.com/JonathanSalwan/Triton/issues/831#ref-commit-ab37265).  At some point I need to break open the PowerPC ISA and make changes to the ppc32.spec file and several other files. \o/

**Comment by JonathanSalwan** [comment]
> I forked Triton's master repository

Never use `master` as base. Please use `dev-ppc`.

> I'm sort of following along each commit in thread #831.

For a better overview, you should use this https://github.com/JonathanSalwan/Triton/compare/dev-v0.8...dev-arm32 instead of commits in thread #831. 

> and several other files. \o/

Please don't touch as much as possible files which are not explicitly related to PPC (such as SymbolicEngine, etc.).

Some advice to help me during review:

* Please use the same coding style (space, var name, mangling, new lines, etc.)
* Please ask before modifying core files
* Please don't add complex C++ magic voodoo stuffs
* Please add a lot of comments (almost at each line)
* Please keep up-to-date the Doxygen documentation
* Please for your instruction semantics add a lot of unittests (I will never merge a PR without unittest)
* Please don't add all instruction semantics before I approved the PPC specifications (add only one or two instructions)
* Please keep me up-to-date before making a PR of 10k lines...

The main issue I see adding PPC is the big endian which necessitate core files modification... Btw, I still thinking that it's a bit early for adding new architectures. Our core engines will probably change (ex. #473, #753, #806) and it will probably imply that we will have to modify all *Semantics.cpp files... More we have architectures more it will boring to update them :). Our philosophy is to provide few but solid features instead of a lot of unfinished features.

**Comment by jpc0016** [comment]
Thanks for the feedback Jonathan!  

I figured it was still early since I saw versions starting with '0.x' and not '1.x'  I will make fork changes soon and recommit the changes I made.  I agree on standardizing your C++ style across all files.  There were no additional unique lines; basically changing ARM nomenclature to PPC.  Because Triton has not emerged as version 1.x release, I will reserve all PPC changes and keep up to date on Triton progress!
